Tweaking App Review Prompt

sendhil opened the pull request:
Tweaking the App Review prompt to allow for more granularity. Also added ability for app review prompt to be disabled remotely on a feature by feature basis or for the entire app.

cc @diegoreymendez

diegoreymendez commented on WordPress/Classes/Utility/Ratings/AppRatingUtility.h:
```diff
@@ -1,15 +1,105 @@
-// Based loosely on Appirater (https://github.com/arashpayan/appirater)
-
+/**
+ This class will help track whether or not a user should be prompted for an app review. This class is loosely based on Appirater (https://github.com/arashpayan/appirater)
```
This is pretty minor, but if you could break comment lines to fit the default size (I think it's 80 chars) it would help for reading them in github.  Documentation generators can deal with hard breaks so it shouldn't be a problem on that front.

diegoreymendez commented on WordPress/Classes/Utility/Ratings/AppRatingUtility.m:
```diff
@@ -53,23 +112,57 @@ + (void)initializeForVersion:(NSString *)version
+                    // We haven't skipped enough versions, skip this one
+                    numberOfVersionsSkippedPrompting++;
+                    [userDefaults setInteger:numberOfVersionsSkippedPrompting forKey:AppRatingNumberOfVersionsSkippedPrompting];
+                    [userDefaults setBool:YES forKey:AppRatingRatedCurrentVersion];
```
Should this be `NO` instead?

diegoreymendez replied:
I just realized that the reason this is set to YES is probably to trigger the `interactedWithAppReviewPromptInPreviousVersion` condition above on the next version of the app.  Is this correct?

If so, I'd consider leaving this as NO (since the user didn't really rate the current version, and it could be hard to read for other developers, and instead replace the following line above:

`if (interactedWithAppReviewPromptInPreviousVersion) {`

...with something like...

`if ([userDefaults integerForKey:AppRatingNumberOfVersionsToSkipPrompting] > 0) {`

diegoreymendez commented on WordPress/Classes/Utility/Ratings/AppRatingUtility.m:
```diff
@@ -1,38 +1,97 @@
+    }];
+}
+
++ (void)checkIfAppReviewPromptsHaveBeenDisabled:(void (^)(void))success failure:(void (^)(void))failure;
```
Again, minor... but if you:

`typedef void (^AppRatingCompletionBlock)();`

You can then:

`+ (void)checkIfAppReviewPromptsHaveBeenDisabled:(AppRatingCompletionBlock)success failure:(AppRatingCompletionBlock)failure;

Which also lets callers use:

`AppRatingCompletionBlock success = ^void() {...}`

Feel free to opt out of this if you don't like it though!

diegoreymendez commented:
First pass is complete.  I run some manual tests on the app, run the unit tests, and checked the code as thoroughly as I could.  Nice work!  I especially liked the extra documentation you added and the unit test coverage.

diegoreymendez commented:
Let me know what you want to do with my comments above.

sendhil commented:
Hey @diegoreymendez - addressed your comments - see https://github.com/wordpress-mobile/WordPress-iOS/commit/20ddaaba9af959c6b0f34d4728cf95c72143723d for the big one.

diegoreymendez commented:
- Code looks fine to me.
- Unit tests run just fine (with the exception of some unrelated tests).
- The feature worked fine during my manual tests.

:shipit:
